KAFKA-16341: fix the LogValidator for non-compressed type

johnnychhsu opened the pull request:
## Context
Previously in the LogValidator, the `offsetOfMaxTimestamp` depends on two parameter check
1. timestampType
2. batch.toMagic

If the `timestampType` is `LOG_APPEND_TIME`, and the `toMagic` is larger or equals to `MAGIC_VALUE_V2`, then we assign `offsetCounter.value - 1` to it, otherwise keep the initial value.

However, in [KAFKA-14477](https://issues.apache.org/jira/browse/KAFKA-14477), this was changed, which modify `offsetOfMaxTimestamp` no matter whether it's `LOG_APPEND_TIME` or not, but only verify the `toMagic`, which led to this issue. More details can be found in [KAFKA-16310](https://issues.apache.org/jira/browse/KAFKA-16310)

Besides, in [KIP-734](https://cwiki.apache.org/confluence/display/KAFKA/KIP-734%3A+Improve+AdminClient.listOffsets+to+return+timestamp+and+offset+for+the+record+with+the+largest+timestamp), we have enables `MAX_TIMESTAMP` of listOffset, where the name `shallowOffsetOfMaxTimestamp` become confusing since we actually get the offset of the max timestamp at the record level, not in the batch level.

## Solution
- Fix the verifying logic. If it's `LOG_APPEND_TIME`, we choose the offset of the first record. Else, we choose the record with the maxTimeStamp. 
- rename the shallowOffsetOfMaxTimestamp to offsetOfMaxTimestamp

## Test
test with
```
./gradlew clean core:test --tests kafka.log.LogValidatorTest
./gradlew clean client:test --tests org.apache.kafka.common.record.MemoryRecordsBuilderTest
./gradlew clean core:test --tests kafka.admin.ListOffsetsIntegrationTest
./gradlew clean core:test --tests kafka.log.LogSegmentTest
./gradlew clean core:test --tests kafka.log.LocalLogTest 
./gradlew clean client:test --tests org.apache.kafka.common.record.MemoryRecordsBuilderTest
```
and it passed

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

Commit:
fix the LogValidator for non-compredded type

Commit:
remove test for compressed case

chia7712 commented:
The fix is perfect. Please rewrite the test according to #15474. Putting all test cases together is more readable.

johnnychhsu commented:
thanks for the quick review @chia7712 , sure let me address that

junrao commented on storage/src/main/java/org/apache/kafka/storage/internals/log/LogValidator.java:
```diff
@@ -293,11 +293,11 @@ public ValidationResult assignOffsetsNonCompressed(LongRef offsetCounter,
+            if (toMagic >= RecordBatch.MAGIC_VALUE_V2) {
+                offsetOfMaxTimestamp = offsetCounter.value - 1;
+            } else {
+                offsetOfMaxTimestamp = initialOffset;
```
We pass `offsetOfMaxTimestamp` to `ValidationResult.shallowOffsetOfMaxTimestampMs`. Using shallow offset used to be ok.  But when we added the `MAX_TIMESTAMP` support for ListOffset, it becomes important for offsetOfMaxTimestamp to be at the exact record level, instead of the shallow batch level. So, we need to rename `ValidationResult.shallowOffsetOfMaxTimestampMs` properly to make it clear. 

Also, there are other places like `LogSegment.append()` where we use the name `shallowOffsetOfMaxTimestamp`. We will need to rename it too to be clear.

johnnychhsu replied:
thanks for the comment! 
yes the naming is confusing, although it does represent the latest offset but it looks like it's only the offset of one batch, instead of the record. Let me address this.

johnnychhsu replied:
I have created a ticket [KAFKA-16380](https://issues.apache.org/jira/browse/KAFKA-16380) to track the changes of shallowOffsetOfMaxTimestamp for LogSegment

junrao reviewed (commented):
@johnnychhsu : Thanks for the PR. Left a comment below.

junrao commented on core/src/test/scala/integration/kafka/api/OffsetOfMaxTimestampTest.scala:
```diff
@@ -0,0 +1,77 @@
+        val time: Long = 10000
+        producer.send(new ProducerRecord[String, String](topicName, 0, time + 100, null, "val20"))
+        producer.send(new ProducerRecord[String, String](topicName, 0, time + 400, null, "val15"))
+        producer.send(new ProducerRecord[String, String](topicName, 0, time + 250, null, "val15"))
```
How do we make sure that the 3 records are in the same batch?

chia7712 replied:
https://github.com/apache/kafka/pull/15474#discussion_r1518055609 can address that.

jolshan replied:
Do we need a separate test for this? I mentioned how to test here: https://github.com/apache/kafka/pull/15474#discussion_r1516625703

jolshan replied:
For the non-compression case, we can just make the record size 10 B (rather than 10KB). 
There is also some discussion about how to control the batch size via flush

chia7712 replied:
@jolshan I prefer to have following test: https://github.com/apache/kafka/pull/15474#discussion_r1520206161

junrao commented on storage/src/main/java/org/apache/kafka/storage/internals/log/LogValidator.java:
```diff
@@ -293,11 +293,11 @@ public ValidationResult assignOffsetsNonCompressed(LongRef offsetCounter,
-        if (toMagic >= RecordBatch.MAGIC_VALUE_V2) {
-            offsetOfMaxTimestamp = offsetCounter.value - 1;
+            if (toMagic >= RecordBatch.MAGIC_VALUE_V2) {
+                offsetOfMaxTimestamp = offsetCounter.value - 1;
```
Hmm, we should be consistent among different magic values, right? It seems the semantic for MAX_TIMESTAMP is the first offset with the max timestamp.

johnnychhsu replied:
thanks for the comment!
this is the same behaviour before the previous change affected the results. It differentiated the behaviour based on the magic value. I think the reason is because


> /**
> A record batch is a container for records. In old versions of the record format (versions 0 and 1),
> a batch consisted always of a single record if no compression was enabled, but could contain
> many records otherwise. Newer versions (magic versions 2 and above) will generally contain many records
> regardless of compression.
>*/
> public interface RecordBatch extends Iterable<Record>

So version before magic_value=2 doesn't matter.

junrao replied:
@johnnychhsu : Yes, that was the original code. But the issue is that that code was no longer correct when [KIP-734](https://cwiki.apache.org/confluence/display/KAFKA/KIP-734%3A+Improve+AdminClient.listOffsets+to+return+timestamp+and+offset+for+the+record+with+the+largest+timestamp) was added. Without KIP-734, it's ok to set offsetOfMaxTimestamp to the last offset for magic V2. However, with KIP-734, it's important to set offsetOfMaxTimestamp to the first offset for both magic V1 and V2. It seems that we missed that part when adding KIP-734. So, I think the logic should be if `timestampType == TimestampType.LOG_APPEND_TIME`, we always set `offsetOfMaxTimestamp` to `initialOffset`.

chia7712 replied:
It seems KIP-734 does not define the case we are discussing - return the initial/last offset when there are multi-records having same "max timestamp".  

So how we deal with it? Is it a good practice to discuss compatibility? Or we deal with it as "bug fix" and we add enough document to explain the "correct behavior" ?

I prefer the consistent behavior. Hence, it seems to me that adding document to `OffsetSpec.MaxTimestampSpec` is enough here. The use cases which depends on previous behavior (last offset) should be re-written for this bug fix ( and behavior change).

showuon replied:
> I prefer the consistent behavior. Hence, it seems to me that adding document to OffsetSpec.MaxTimestampSpec is enough here. The use cases which depends on previous behavior (last offset) should be re-written for this bug fix ( and behavior change).

+1 from me.

junrao replied:
> It seems KIP-734 does not define the case we are discussing - return the initial/last offset when there are multi-records having same "max timestamp".

Yes, I agree that we need to be consistent. Using the initial offset makes the most intuitive sense to me. We could update KIP-734 with this clarification and also send it to the original vote thread to make sure there is no objection.

showuon commented:
Please also address this comment in this PR: https://github.com/apache/kafka/pull/15474#discussion_r1520262997 . Thanks.

chia7712 commented:
@johnnychhsu Could you please rebase code and address following comments.

1. update KIP-734 according to https://github.com/apache/kafka/pull/15476#discussion_r1523677272

2. complete the tests according to https://github.com/apache/kafka/pull/15474#discussion_r1524067155

3. remove the shallow part in RecordsInfo.shallowOffsetOfMaxTimestamp according to https://github.com/apache/kafka/pull/15474#discussion_r1520262997

Sorry that we bring some extra works to this PR, and thank to you for taking over this PR :)

chia7712 commented on core/src/test/scala/integration/kafka/api/OffsetOfMaxTimestampTest.scala:
```diff
@@ -0,0 +1,77 @@
+
+import java.util.{Collections, Properties}
+
+class OffsetOfMaxTimestampTest extends IntegrationTestHarness {
```
please remove this test and enhance the test `ListOffsetsIntegrationTest`.

johnnychhsu replied:
thanks for the reminder!
I have removed this test and add the test case in ListOffsetsIntegrationTest

Commit:
address comment

Commit:
remove scala test

Commit:
update comments

johnnychhsu commented:
thanks @junrao @chia7712 @showuon for the suggestions and comments!
I have updated the following:
- rename the ValidationResult.shallowOffsetOfMaxTimestamp to ValidationResult.recordOffsetOfMaxTimestampMs
- rename the RecordsInfo.shallowOffsetOfMaxTimestamp to RecordsInfo.recordOffsetOfMaxTimestampMs
- modify some comments to replace the "shallow" wording
- added tests in ListOffsetsIntegrationTest.testThreeRecordsInSeparateBatch

junrao commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java:
```diff
@@ -209,7 +209,7 @@ private static FilterResult filterTo(TopicPartition partition, Iterable<MutableR
 
                         MemoryRecordsBuilder.RecordsInfo info = builder.info();
-                        filterResult.updateRetainedBatchMetadata(info.maxTimestamp, info.shallowOffsetOfMaxTimestamp,
+                        filterResult.updateRetainedBatchMetadata(info.maxTimestamp, info.recordOffsetOfMaxTimestamp,
```
There are still a few places with `shallowOffsetOfMaxTimestamp`. We need to fix them all.

`FilterResult` and `updateRetainedBatchMetadata()` in `MemoryRecords`.
`append()` in `LocalLog` 
`append()` in `LogSegment`

chia7712 replied:
It seems `OffsetOfMaxTimestamp` is simpler and readable. I like the naming relation between `info.maxTimestamp` and `info.offsetOfmaxTimestamp` :)

johnnychhsu replied:
sounds good to me : )

junrao commented on core/src/test/scala/integration/kafka/admin/ListOffsetsIntegrationTest.scala:
```diff
@@ -82,6 +82,13 @@ class ListOffsetsIntegrationTest extends KafkaServerTestHarness {
+    props.setProperty(TopicConfig.MESSAGE_TIMESTAMP_TYPE_CONFIG, "LogAppendTime")
+    createTopicWithConfig(topicNameWithCustomConfigs, props)
+    produceMessagesInSeparateBatch(topic = topicNameWithCustomConfigs)
+    verifyListOffsets(topic = topicNameWithCustomConfigs, 2)
```
It's fine to add this test, but it doesn't require the fix in LogValidator in this PR, right? If each batch has only a single record, the offset for maxTimestamp is always the same between logAppend and create time. 

It seems that we need to add a test case that includes multiple records in a non-compressed batch.

johnnychhsu replied:
thanks for the comment!
yes, we should have a test for non-compressed type, where multiple records are sent in one batch, then test the offset to verify the fix for LogValidator. Let me updated this.

chia7712 replied:
it would be nice to add comments for this case.

johnnychhsu replied:
agree, let me add that. Thanks for the suggestion!

Commit:
fix failed test

johnnychhsu commented:
test with 
```
 ./gradlew clean core:test --tests kafka.log.LogValidatorTest
./gradlew clean client:test --tests org.apache.kafka.common.record.MemoryRecordsBuilderTest
./gradlew clean core:test --tests kafka.admin.ListOffsetsIntegrationTest
```
and passed

johnnychhsu commented on core/src/test/scala/unit/kafka/log/LogValidatorTest.scala:
```diff
@@ -404,14 +404,8 @@ class LogValidatorTest {
     assertEquals(now + 1, validatingResults.maxTimestampMs,
       s"Max timestamp should be ${now + 1}")
 
-    val expectedShallowOffsetOfMaxTimestamp = if (magic >= RecordVersion.V2.value) {
```
if it's LOG_APPEND_TIME, it should be the first record. If not, then we should choose the correct one

junrao commented on core/src/test/scala/unit/kafka/log/LogValidatorTest.scala:
```diff
@@ -173,9 +173,9 @@ class LogValidatorTest {
-    // we index from last offset in version 2 instead of base offset
-    val expectedMaxTimestampOffset = if (magic >= RecordBatch.MAGIC_VALUE_V2) 2 else 0
-    assertEquals(expectedMaxTimestampOffset, validatedResults.shallowOffsetOfMaxTimestampMs,
+    // If it's LOG_APPEND_TIME, the offset will be the first offset of the record
```
the first offset of the record => the offset of the first record

johnnychhsu replied:
thanks for the comment! 
you are right, let me update this sentence

junrao reviewed (commented):
@johnnychhsu : Thanks for the updated PR. A few more comments.

Commit:
add test for noncompressed in on batch

chia7712 commented on core/src/test/scala/integration/kafka/admin/ListOffsetsIntegrationTest.scala:
```diff
@@ -77,11 +77,34 @@ class ListOffsetsIntegrationTest extends KafkaServerTestHarness {
+
   @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumName)
   @ValueSource(strings = Array("zk", "kraft"))
   def testThreeRecordsInSeparateBatch(quorum: String): Unit = {
```
Should we rename it to `testThreeNonCompressedRecordsInSeparateBatch`?

johnnychhsu replied:
thanks for the suggestions! totally agree, let me change it

chia7712 reviewed (commented):
@johnnychhsu thanks for updating code. a couple of comments left.

Commit:
addressing comments

johnnychhsu commented:
test with
```
./gradlew clean core:test --tests kafka.log.LogSegmentTest
./gradlew clean core:test --tests kafka.log.LocalLogTest 
```
and it passed

johnnychhsu commented on core/src/test/scala/integration/kafka/admin/ListOffsetsIntegrationTest.scala:
```diff
@@ -79,9 +79,34 @@ class ListOffsetsIntegrationTest extends KafkaServerTestHarness {
+    props.setProperty(TopicConfig.MESSAGE_TIMESTAMP_TYPE_CONFIG, "LogAppendTime")
+    createTopicWithConfig(topicNameWithCustomConfigs, props)
+    produceMessagesInOneBatch(topic=topicNameWithCustomConfigs)
+    // In LogAppendTime's case, if the timestamps are the same, we choose the offset of the first record
```
this is the difference explanation compared to previous behaviour

chia7712 commented:
@johnnychhsu Could you check the failed tests?

johnnychhsu commented:
@chia7712 thanks for the reminder.

Below are the failed tests from the build 
> [Build / JDK 17 and Scala 2.13 / kafka.api.CustomQuotaCallbackTest.testCustomQuotaCallback()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/kafka.api/CustomQuotaCallbackTest/Build___JDK_17_and_Scala_2_13___testCustomQuotaCallback__/)
> [Build / JDK 17 and Scala 2.13 / kafka.api.DelegationTokenEndToEndAuthorizationWithOwnerTest."testNoDescribeProduceOrConsumeWithoutTopicDescribeAcl(String, boolean).quorum=kraft, isIdempotenceEnabled=false"](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/kafka.api/DelegationTokenEndToEndAuthorizationWithOwnerTest/Build___JDK_17_and_Scala_2_13____testNoDescribeProduceOrConsumeWithoutTopicDescribeAcl_String__boolean__quorum_kraft__isIdempotenceEnabled_false_/)
> [Build / JDK 17 and Scala 2.13 / kafka.server.LogDirFailureTest.testIOExceptionDuringCheckpoint(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/kafka.server/LogDirFailureTest/Build___JDK_17_and_Scala_2_13___testIOExceptionDuringCheckpoint_String__quorum_kraft/)
> [Build / JDK 17 and Scala 2.13 / kafka.server.ReplicaManagerTest.testRemoteFetchExpiresPerSecMetric()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/kafka.server/ReplicaManagerTest/Build___JDK_17_and_Scala_2_13___testRemoteFetchExpiresPerSecMetric__/)
> [Build / JDK 11 and Scala 2.13 / org.apache.kafka.common.network.SslTransportLayerTest."testUngracefulRemoteCloseDuringHandshakeRead(Args).tlsProtocol=TLSv1.2, useInlinePem=true"](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/org.apache.kafka.common.network/SslTransportLayerTest/Build___JDK_11_and_Scala_2_13____testUngracefulRemoteCloseDuringHandshakeRead_Args__tlsProtocol_TLSv1_2__useInlinePem_true_/)
> [Build / JDK 11 and Scala 2.13 / kafka.api.ConsumerBounceTest.testConsumptionWithBrokerFailures()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/kafka.api/ConsumerBounceTest/Build___JDK_11_and_Scala_2_13___testConsumptionWithBrokerFailures__/)
> [Build / JDK 11 and Scala 2.13 / kafka.api.PlaintextConsumerTest.testExpandingTopicSubscriptions(String, String).quorum=kraft+kip848.groupProtocol=consumer](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/kafka.api/PlaintextConsumerTest/Build___JDK_11_and_Scala_2_13___testExpandingTopicSubscriptions_String__String__quorum_kraft_kip848_groupProtocol_consumer/)
> [Build / JDK 11 and Scala 2.13 / kafka.network.DynamicConnectionQuotaTest.testDynamicListenerConnectionCreationRateQuota(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/kafka.network/DynamicConnectionQuotaTest/Build___JDK_11_and_Scala_2_13___testDynamicListenerConnectionCreationRateQuota_String__quorum_kraft/)
> [Build / JDK 11 and Scala 2.13 / kafka.server.ControllerRegistrationManagerTest.testWrongIncarnationId()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/kafka.server/ControllerRegistrationManagerTest/Build___JDK_11_and_Scala_2_13___testWrongIncarnationId__/)
> [Build / JDK 11 and Scala 2.13 / kafka.server.LogDirFailureTest.testIOExceptionDuringLogRoll(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/kafka.server/LogDirFailureTest/Build___JDK_11_and_Scala_2_13___testIOExceptionDuringLogRoll_String__quorum_kraft/)
> [Build / JDK 11 and Scala 2.13 / kafka.server.ReplicaManagerTest.testRemoteFetchExpiresPerSecMetric()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/kafka.server/ReplicaManagerTest/Build___JDK_11_and_Scala_2_13___testRemoteFetchExpiresPerSecMetric__/)
> [Build / JDK 11 and Scala 2.13 / kafka.zk.ZkMigrationIntegrationTest.testNewAndChangedTopicsInDualWrite [1] Type=ZK, MetadataVersion=3.4-IV0, Security=PLAINTEXT](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/kafka.zk/ZkMigrationIntegrationTest/Build___JDK_11_and_Scala_2_13___testNewAndChangedTopicsInDualWrite__1__Type_ZK__MetadataVersion_3_4_IV0__Security_PLAINTEXT/)
> [Build / JDK 11 and Scala 2.13 / org.apache.kafka.tools.MetadataQuorumCommandTest.testDescribeQuorumReplicationSuccessful [5] Type=Raft-Combined, MetadataVersion=3.8-IV0, Security=PLAINTEXT](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/org.apache.kafka.tools/MetadataQuorumCommandTest/Build___JDK_11_and_Scala_2_13___testDescribeQuorumReplicationSuccessful__5__Type_Raft_Combined__MetadataVersion_3_8_IV0__Security_PLAINTEXT/)
> [Build / JDK 11 and Scala 2.13 / org.apache.kafka.trogdor.coordinator.CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/org.apache.kafka.trogdor.coordinator/CoordinatorTest/Build___JDK_11_and_Scala_2_13___testTaskRequestWithOldStartMsGetsUpdated__/)
> [Build / JDK 8 and Scala 2.12 / kafka.server.LogDirFailureTest.testIOExceptionDuringLogRoll(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/kafka.server/LogDirFailureTest/Build___JDK_8_and_Scala_2_12___testIOExceptionDuringLogRoll_String__quorum_kraft/)
> [Build / JDK 8 and Scala 2.12 / kafka.server.ReplicaManagerTest.testRemoteFetchExpiresPerSecMetric()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/kafka.server/ReplicaManagerTest/Build___JDK_8_and_Scala_2_12___testRemoteFetchExpiresPerSecMetric__/)
> [Build / JDK 8 and Scala 2.12 / org.apache.kafka.controller.QuorumControllerTest.testBootstrapZkMigrationRecord()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_8_and_Scala_2_12___testBootstrapZkMigrationRecord__/)
> [Build / JDK 8 and Scala 2.12 / org.apache.kafka.tools.MetadataQuorumCommandTest.testDescribeQuorumStatusSuccessful [2] Type=Raft-Isolated, MetadataVersion=3.8-IV0, Security=PLAINTEXT](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/org.apache.kafka.tools/MetadataQuorumCommandTest/Build___JDK_8_and_Scala_2_12___testDescribeQuorumStatusSuccessful__2__Type_Raft_Isolated__MetadataVersion_3_8_IV0__Security_PLAINTEXT/)
> [Build / JDK 8 and Scala 2.12 / org.apache.kafka.tools.MetadataQuorumCommandTest.testDescribeQuorumStatusSuccessful [6] Type=Raft-Isolated, MetadataVersion=3.8-IV0, Security=PLAINTEXT](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/org.apache.kafka.tools/MetadataQuorumCommandTest/Build___JDK_8_and_Scala_2_12___testDescribeQuorumStatusSuccessful__6__Type_Raft_Isolated__MetadataVersion_3_8_IV0__Security_PLAINTEXT/)
> [Build / JDK 21 and Scala 2.13 / org.apache.kafka.connect.integration.ExactlyOnceSourceIntegrationTest.testSeparateOffsetsTopic](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/org.apache.kafka.connect.integration/ExactlyOnceSourceIntegrationTest/Build___JDK_21_and_Scala_2_13___testSeparateOffsetsTopic/)
> [Build / JDK 21 and Scala 2.13 / kafka.api.ConsumerBounceTest.testConsumptionWithBrokerFailures()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/kafka.api/ConsumerBounceTest/Build___JDK_21_and_Scala_2_13___testConsumptionWithBrokerFailures__/)
> [Build / JDK 21 and Scala 2.13 / kafka.api.SaslPlainPlaintextConsumerTest.testCoordinatorFailover(String, String).quorum=kraft.groupProtocol=classic](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/kafka.api/SaslPlainPlaintextConsumerTest/Build___JDK_21_and_Scala_2_13___testCoordinatorFailover_String__String__quorum_kraft_groupProtocol_classic/)
> [Build / JDK 21 and Scala 2.13 / kafka.api.SaslPlainPlaintextConsumerTest.testCoordinatorFailover(String, String).quorum=kraft+kip848.groupProtocol=classic](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/kafka.api/SaslPlainPlaintextConsumerTest/Build___JDK_21_and_Scala_2_13___testCoordinatorFailover_String__String__quorum_kraft_kip848_groupProtocol_classic/)
> [Build / JDK 21 and Scala 2.13 / kafka.api.SaslPlaintextConsumerTest.testCoordinatorFailover(String, String).quorum=kraft.groupProtocol=classic](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/kafka.api/SaslPlaintextConsumerTest/Build___JDK_21_and_Scala_2_13___testCoordinatorFailover_String__String__quorum_kraft_groupProtocol_classic/)
> [Build / JDK 21 and Scala 2.13 / kafka.coordinator.transaction.ProducerIdManagerTest."testUnrecoverableErrors(Errors).error=INVALID_REQUEST"](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/kafka.coordinator.transaction/ProducerIdManagerTest/Build___JDK_21_and_Scala_2_13____testUnrecoverableErrors_Errors__error_INVALID_REQUEST_/)
> [Build / JDK 21 and Scala 2.13 / kafka.server.LogDirFailureTest.testIOExceptionDuringLogRoll(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/kafka.server/LogDirFailureTest/Build___JDK_21_and_Scala_2_13___testIOExceptionDuringLogRoll_String__quorum_kraft/)
> [Build / JDK 21 and Scala 2.13 / kafka.server.LogDirFailureTest.testIOExceptionDuringLogRoll(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/kafka.server/LogDirFailureTest/Build___JDK_21_and_Scala_2_13___testIOExceptionDuringLogRoll_String__quorum_kraft_2/)
> [Build / JDK 21 and Scala 2.13 / kafka.server.ReplicaManagerTest.testRemoteFetchExpiresPerSecMetric()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/kafka.server/ReplicaManagerTest/Build___JDK_21_and_Scala_2_13___testRemoteFetchExpiresPerSecMetric__/)
> [Build / JDK 21 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testFenceMultipleBrokers()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15476/10/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_21_and_Scala_2_13___testFenceMultipleBrokers__/)

test with
```
 ./gradlew cleanTest core:test --tests CustomQuotaCallbackTest --tests LogDirFailureTest --tests ReplicaManagerTest --tests ConsumerBounceTest --tests PlaintextConsumerTest --tests DynamicConnectionQuotaTest --tests ControllerRegistrationManagerTest --tests ZkMigrationIntegrationTest --tests SaslPlainPlaintextConsumerTest --tests ProducerIdManagerTest clients:test --tests SslTransportLayerTest tools:test --tests MetadataQuorumCommandTest connect:test --tests ExactlyOnceSourceIntegrationTest
```
and it passed

as for `DelegationTokenEndToEndAuthorizationWithOwnerTest`, which looks like a flaky test [KAFKA-15411](https://issues.apache.org/jira/browse/KAFKA-15411) I tried with 
```
./gradlew cleanTest core:test --tests DelegationTokenEndToEndAuthorizationWithOwnerTest
```
and it succeed

chia7712 reviewed (approved):
LGTM

will merge it tomorrow

showuon commented:
I will take a look later.

showuon commented on core/src/test/scala/integration/kafka/admin/ListOffsetsIntegrationTest.scala:
```diff
@@ -79,9 +79,34 @@ class ListOffsetsIntegrationTest extends KafkaServerTestHarness {
+    produceMessagesInOneBatch()
+    verifyListOffsets()
+
+    // test LogAppendTime case
+    val props: Properties = new Properties()
+    props.setProperty(TopicConfig.MESSAGE_TIMESTAMP_TYPE_CONFIG, "LogAppendTime")
+    createTopicWithConfig(topicNameWithCustomConfigs, props)
+    produceMessagesInOneBatch(topic=topicNameWithCustomConfigs)
+    // In LogAppendTime's case, if the timestamps are the same, we choose the offset of the first record
```
These codes to verify the logAppendTime has duplicated quite a lot in this test suite. Could we extract the into another method?

johnnychhsu replied:
totally agree, let me check this, thanks for the suggestion!

johnnychhsu replied:
@showuon just updated, thanks for the review!

showuon reviewed (commented):
Thanks for the PR. LGTM! Just one minor comment.

Commit:
refactor ListOffsetsIntegrationTest

junrao reviewed (approved):
@johnnychhsu : Thanks for the updated PR. LGTM

The jdk 8 tests didn't complete. Perhaps we could run it again.

showuon reviewed (approved):
LGTM!

chia7712 committed:
Merge branch 'trunk' into KAFKA-16341/fix_uncompressed_records

chia7712 commented:
I have rebase the PR to trigger QA again, and I will merge it if no related failed tests.

chia7712 commented:
JDK 11 gets cancel. It should be fine as it gets test in previous QA

I run the failed tests on my local:
```sh
./gradlew cleanTest :tools:test --tests MetadataQuorumCommandTest.testDescribeQuorumStatusSuccessful --tests MetadataQuorumCommandTest.testDescribeQuorumReplicationSuccessful :storage:test --tests TransactionsWithTieredStoreTest.testSendOffsetsToTransactionTimeout --tests TransactionsWithTieredStoreTest.testAbortTransactionTimeout --tests TransactionsWithTieredStoreTest.testCommitTransactionTimeout :connect:runtime:test --tests org.apache.kafka.connect.integration.ExactlyOnceSourceIntegrationTest.testSeparateOffsetsTopic --tests org.apache.kafka.connect.integration.OffsetsApiIntegrationTest.testAlterSinkConnectorOffsetsOverriddenConsumerGroupId :metadata:test --tests QuorumControllerTest.testBootstrapZkMigrationRecord --tests QuorumControllerTest.testFenceMultipleBrokers --tests QuorumControllerTest.testBalancePartitionLeaders :trogdor:test --tests CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated :connect:mirror:test --tests IdentityReplicationIntegrationTest.testReplicateSourceDefault --tests MirrorConnectorsIntegrationBaseTest.testReplicateSourceDefault --tests MirrorConnectorsIntegrationBaseTest.testSyncTopicConfigs --tests MirrorConnectorsIntegrationExactlyOnceTest.testReplication --tests MirrorConnectorsWithCustomForwardingAdminIntegrationTest.testSyncTopicConfigs --tests MirrorConnectorsIntegrationTransactionsTest.testReplicateSourceDefault --tests MirrorConnectorsIntegrationSSLTest.testSyncTopicConfigs :core:test --tests DelegationTokenEndToEndAuthorizationWithOwnerTest.testNoConsumeWithDescribeAclViaSubscribe --tests DelegationTokenEndToEndAuthorizationWithOwnerTest.testDescribeTokenForOtherUserPasses --tests ConsumerBounceTest.testClose --tests ConsumerBounceTest.testConsumptionWithBrokerFailures --tests PlaintextAdminIntegrationTest.testElectPreferredLeaders --tests LogDirFailureTest.testIOExceptionDuringLogRoll --tests LogDirFailureTest.testIOExceptionDuringCheckpoint --tests ReplicaManagerTest.testRemoteFetchExpiresPerSecMetric --tests LogRecoveryTest.testHWCheckpointWithFailuresMultipleLogSegments
```
all pass, so I'm about to merge it

chia7712 commented:
@johnnychhsu thanks for this contribution! please complete the follow-up:

1. 3.6 branch has many conflicts so please file a PR for branch 3.6
2. update KIP-734 for the offset of max timestamp

I have backport to 3.7 manually due to less conflicts.
